### PR TITLE
add interactive editor session help menu

### PR DESCRIPTION
**Pull request by meganrogge**
Part of https://github.com/microsoft/vscode-internalbacklog/issues/4109

Adds a help menu to the input box and also adds a const enum `AccessibilityVerbositySettingId`

added `feature-request` so I remember to add this to a TPI

### CODE REVIEW

File: src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorActions.ts [hunk 1]
```diff
@@ -547,3 +548,21 @@ export class ContractMessageAction extends AbstractInteractiveEditorAction {
 		ctrl.updateExpansionState(false);
 	}
 }
+
+export class AccessibilityHelpEditorAction extends EditorAction {
```
Comment (jrieken) [inline src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorActions.ts:552]: Please use `EditorAction2` has other actions do in this file

File: src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorWidget.ts [hunk 2]
```diff
@@ -461,8 +463,28 @@ export class InteractiveEditorWidget {
 		this._previewDiffEditor.getOriginalEditor().setHiddenAreas(hiddenOriginal.map(lineRangeAsRange), 'diff-hidden');
 		this._previewDiffEditor.getModifiedEditor().setHiddenAreas(hiddenModified.map(lineRangeAsRange), 'diff-hidden');
 		this._previewDiffEditor.revealLine(modifiedLineRange.startLineNumber, ScrollType.Immediate);
-
 		this._onDidChangeHeight.fire();
+		if (this._accessibilityService.isScreenReaderOptimized()) {
+			this._previewDiffEditor.onDidUpdateDiff(() => {
```
Comment (jrieken) [inline src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorWidget.ts:468]: This listener leaks. Also I quite don't understand why this need to be done on event not when setting in diff input.

File: src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorWidget.ts [hunk 3]
```diff
@@ -461,8 +463,28 @@ export class InteractiveEditorWidget {
...
+				const container = this._previewDiffEditor.getContainerDomNode();
+				if (!modified || !original) {
+					return;
+				}
+				// prevent escape from focusing the underlying editors
+				this._previewDiffEditor.getContainerDomNode().classList.add('screen-reader-interactive-editor-widget');
+				// prevent the user from having to tab through the underlying editors
+				modified.querySelector('.previewDiff textarea.inputarea')?.setAttribute('disabled', '');
+				original.querySelector('.previewDiff textarea.inputarea')?.setAttribute('disabled', '');
+				// open the diff review
```
Comment (jrieken) [inline src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorWidget.ts:480]: Reaching into the internals of the widget seems wrong. I don't know if there is real API for this but we cannot do things like, it's too fragile.

File: src/vs/editor/browser/widget/diffReview.ts [hunk 4]
```diff
@@ -178,6 +178,10 @@ export class DiffReview extends Disposable {
 				|| e.equals(KeyCode.Space)
 				|| e.equals(KeyCode.Enter)
 			) {
+				if (this._diffEditor.getContainerDomNode().classList.contains('screen-reader-interactive-editor-widget')) {
+					// We don't want to show the underlying editors when the diff editor is in inline mode
```
Comment (jrieken) [inline src/vs/editor/browser/widget/diffReview.ts:182]: This is a conceptual layer breaker and some real API is needed instead

**Review by jrieken — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorActions.ts (hunk 1 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorActions.ts:552]: Yeah I tried and it didn't work - there was an error being thrown related to the number of lines exceeding max line count, will investigate

File: src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorWidget.ts (hunk 2 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorWidget.ts:468]: I couldn't find the right place for this, but am aware it's wrong/ a hack. Will investigate

File: src/vs/editor/browser/widget/diffReview.ts (hunk 4 above)
Comment (meganrogge) [reply src/vs/editor/browser/widget/diffReview.ts:182]: Ok, will add

File: src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorWidget.ts (hunk 3 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorWidget.ts:480]: Yep I couldn't find real API for this but agree this is a hack

File: src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorWidget.ts (hunk 2 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorWidget.ts:492]: this is needed so the screen reader user doesn't have either tediously navigate the original vs modified editors or invoked f7 every time. we are doing that for them

File: src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorActions.ts [hunk 5]
```diff
@@ -547,3 +548,20 @@ export class ContractMessageAction extends AbstractInteractiveEditorAction {
 		ctrl.updateExpansionState(false);
 	}
 }
+
+export class AccessibilityHelpEditorAction extends EditorAction2 {
+	constructor() {
+		super({
+			id: 'interactiveSession.action.accessibilityHelpEditor',
```
Comment (jrieken) [inline src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorActions.ts:555]: nit: maybe better `interactiveEditor.accessibilityHelpEditor`

File: src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorActions.ts [hunk 6]
```diff
@@ -547,3 +548,20 @@ export class ContractMessageAction extends AbstractInteractiveEditorAction {
...
+			title: localize('actions.interactiveSession.accessibiltyHelpEditor', "Interactive Session Editor Accessibility Help"),
+			precondition: ContextKeyExpr.has('interactiveEditorFocused'),
```
Comment (jrieken) [inline src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorActions.ts:557]: better use `CTX_INTERACTIVE_EDITOR_FOCUSED` like [here](https://github.com/microsoft/vscode/blob/17db88a3073b2ac067142685bf056fff1e3586c9/src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorActions.ts#L110)

File: src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorWidget.ts [hunk 7]
```diff
@@ -289,6 +304,18 @@ export class InteractiveEditorWidget {
...
 
+	private _updateAriaLabel(): void {
+		if (!this._accessibilityService.isScreenReaderOptimized()) {
+			return;
+		}
+		if (this._configurationService.getValue<boolean>(AccessibilityVerbositySettingId.InteractiveEditor)) {
+			const kbLabel = this._keybindingService.lookupKeybinding('interactiveSession.action.accessibilityHelpEditor')?.getLabel();
+			_inputEditorOptions.ariaLabel = kbLabel ? localize('interactiveSession.action.accessibilityHelpEditor', "Interactive Editor Input, Use {0} for Interactive Editor Accessibility Help.", kbLabel) : localize('interactiveSessionInput.accessibilityHelpNoKb', "Interactive Editor Input, Run the Interactive Editor Accessibility Help command for more information.");
+		} else {
+			_inputEditorOptions.ariaLabel = defaultAriaLabel;
```
Comment (jrieken) [inline src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorWidget.ts:315]: I believe changing `ariaLabel` doesn't have an effect once the editor has been created. Try `_inputEditor.updateOptions({ ariaLabel: ....})`

File: src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorActions.ts [hunk 8]
```diff
@@ -547,3 +548,20 @@ export class ContractMessageAction extends AbstractInteractiveEditorAction {
 		ctrl.updateExpansionState(false);
 	}
 }
+
+export class AccessibilityHelpEditorAction extends EditorAction2 {
+	constructor() {
+		super({
+			id: 'interactiveSession.action.accessibilityHelpEditor',
+			title: localize('actions.interactiveSession.accessibiltyHelpEditor', "Interactive Session Editor Accessibility Help"),
```
Comment (jrieken) [inline src/vs/workbench/contrib/interactiveEditor/browser/interactiveEditorActions.ts:556]: Maybe use `AbstractInteractiveEditorAction.category` and keep the title shorter and/or extend from `AbstractInteractiveEditorAction`

**Review by jrieken — DISMISSED** [review]
Thanks for doing this. I mostly reviewed the interactive editor part and @roblourens might wanna look at chat parts

**Comment by meganrogge** [comment]
fyi @roblourens nothing changed with the chat components - they were just moved around a bit because the editor reuses some of that code

File: src/vs/workbench/contrib/chat/browser/actions/chatAccessibilityHelp.ts [hunk 9]
```diff
@@ -22,3 +54,40 @@ function descriptionForCommand(commandId: string, msg: string, noKbMsg: string,
...
+	}
+
+	const cachedInput = inputEditor.getValue();
+	const cachedPosition = inputEditor.getPosition();
+	inputEditor.getSupportedActions();
+	const helpText = getAccessibilityHelpText(accessor, type, type === 'editor' ? cachedInput : undefined);
+	inputEditor.setValue(helpText);
+	inputEditor.updateOptions({ readOnly: true });
+	inputEditor.focus();
+	const disposable = addStandardDisposableListener(domNode, 'keydown', e => {
```
Comment (roblourens) [inline src/vs/workbench/contrib/chat/browser/actions/chatAccessibilityHelp.ts:78]: Nervous about this, I wonder if it also needs to be disposed when the editor closes or the chat widget goes away.

**Review by roblourens — APPROVED** [review]

**Pull request merged**